pep-9999.rst

```diff
@@ -0,0 +1,868 @@
+
+
+Optimisations and Caching Benefits
+----------------------------------
+
+Immutable objects provide opportunities for optimisation, such as
+structural sharing, memoization, and just-in-time (JIT)
+compilation techniques (specialising for immutable data, e.g.
+fixed shape, fewer barriers, inlining, etc.). Freezing objects can
+allow Python to implement more efficient caching mechanisms and
+enable compiler optimisations that rely on immutability
+assumptions. This PEP will permit such opportunities to go
+beyond today's immutable objects (like ``int``, ``string``) and
+*shallow* immutable objects (``tuple``, ``frozenset``).
+
+
+
+Specification
+=============
+
+Note: our current prototype implementation was authored on-top of
+Python 3.12. To avoid blocking on rebasing on 3.14 to force
+decisions about changes to implementation detail, we are
+circulating this document to discuss the design ideas,
+and some of the unaffected aspects of the implementation.
+
+An outline of the changes that we anticipate are required for
+Python 3.14 is can be found at the `end of the document <Rebasing on Python 3.14>`_.
+
+
+Changes to Python Objects
+-------------------------
+
+Every Python object will have a flag that keeps track of its
+immutability status. Details about the default value of
+this flag is discussed further down in this document. 
+
+The flag can be added without extending the size of the
+Python object header.
+
+
+Implementation of Immutability
+------------------------------
+
+Immutability is enforced through run-time checking. The macro
+``Py_CHECKWRITE(op)`` is inserted on all paths that are guaranteed
+to end up in a write to ``op``. The macro inspects the immutability
+flag in the header of ``op`` and signals an error if the immutability
+flag is set.
+
+A typical use of this check looks like this:
+
+.. code-block:: c
+
+  if (!Py_CHECKWRITE(op)) {        // perform the check
+      PyErr_WriteToImmutable(op);  // raise the error if the check fails
+      return NULL;                 // abort the write
+  }  
+  ... // code that performs the write
+
+
+Writes are common in the CPython code base and the writes lack a
+common "code path" that they pass. To this end, the PEP requires a
+``Py_CHECKWRITE`` call to be inserted and there are several places
+in the CPython code base that are changed as a consequence of this
+PEP. So far we have identified around 70 places in core Python which
+needed a ``Py_CHECKWRITE`` check. Modules in the standard library
+have required somewhere between 5 and 15 checks per module.
+
+
+Backward Compatibility
+======================
+
+This proposal intends to be fully backward compatible, as no existing Python
+code will be affected unless it explicitly calls ``freeze(obj)``.
+Frozen objects will raise errors only when mutation is attempted.
+
+
+Opt-In vs. Opt-Out
+------------------
+
+All pure Python objects can be frozen. However, for types which
```

Comment:
suggested edit: All pure Python objects can be implicitly frozen, provided all their members and their base classes can be frozen.

Reply:
I added it in but dropped the word "implicitly" because that was confusing.